src/Lean/Environment.lean

```diff
@@ -799,6 +894,104 @@ def isSafeDefinition (env : Environment) (declName : Name) : Bool :=
   env.header.moduleNames.findIdx? (· == moduleName)
 
+private structure FindNewDeclsState where
+  postorder : Array AsyncConst := #[]
+  visited   : NameMap Unit := {}
+
+/--
+Starting at `c`, finds constants in `newEnv` but not `oldEnv` and returns them in DAG post-order,
+i.e. in the order they can be replayed onto other environments. New constants that are not from
+`newEnv.asyncConsts` are added with emtpy environment extension state.
+-/
+private partial def findNewAsyncConsts (oldEnv newEnv : Environment) (c : AsyncConst) : Array AsyncConst :=
+  goConst c |>.run {} |>.2.postorder
+where
+  goConst (c : AsyncConst) : StateM FindNewDeclsState Unit := do
+    if oldEnv.contains c.constInfo.name then
+      return
+    match c.constInfo.toConstantInfo with
+    | .defnInfo d | .thmInfo d =>
+        goExpr d.type
+        goExpr d.value
+        modify fun st => { st with postorder := st.postorder.push c, visited := st.visited.insert d.name () }
+    | d => panic! s!"Environment.realizeConst: {d.name} must be definition/theorem"
+  goExpr e := modify (e.foldConsts · fun n st =>
+    if !st.visited.contains n then
+      let c := newEnv.asyncConsts.find? n
+        |>.getD { constInfo := (newEnv.findAsync? n).get!, exts? := none }
+      goConst c |>.run st |>.2
+    else st)
+
+/-- Plumbing function for `Lean.Meta.realizeConst`; see documentation there. -/
+def realizeConst (env : Environment) (forConst : Name) (constName : Name)
+    (realize : Environment → Options → BaseIO (Environment × Dynamic)) :
+    IO (Environment × Dynamic) := do
+  let mut env := env
+  -- find `RealizationContext` for `forConst` in `realizedExternConsts?` or `realizedLocalConsts`
+  let ctx ← if env.checkedWithoutAsync.const2ModIdx.contains forConst then
+    env.realizedExternConsts?.getDM <|
+      throw <| .userError s!"Environment.realizeConst: `realizedExternConsts` is empty"
+  else
+    match env.realizedLocalConsts.find? forConst with
+    | some ctx => pure ctx
+    | none =>
+      throw <| .userError s!"trying to realize {constName} but `enableRealizationsForConst` must be called for '{forConst}' first"
+  let prom ← IO.Promise.new
+  -- ensure `prom` is not left unresolved from stray exceptions
+  BaseIO.toIO do
+    -- atomically check whether we are the first branch to realize `constName`
+    let existingConsts? ← ctx.constsRef.modifyGet fun m => match m.find? constName with
+      | some prom' => (some prom', m)
+      | none       => (none, m.insert constName prom.result!)
+    let (consts, dyn) ← if let some existingConsts := existingConsts? then
+      pure existingConsts.get
+    else
+      -- safety: `RealizationContext` is private
+      let realizeEnv : Environment := unsafe unsafeCast ctx.env
+      let realizeEnv := { realizeEnv with
+        -- allow realizations to recursively realize other constants for `forConst`. Do note that
+        -- this allows for recursive realization of `constName` itself, which will deadlock.
```

Comment:
Would it be hard to guard against that deadlock? Deadlock sounds hard to debug once someone runs into it.

Reply:
We'd need some kind of stack of ongoing realizations...? Tbh it's trivial to debug as a developer if it happens on your machine and this doesn't seem like an issue that is likely to pop up on a user's machine first instead of in our test suite.